Improve parsec validate

matthewrmshin opened the pull request:
(I have kept this branch in my own space for a while now. I am raising this PR for initial discussion. Happy to drop if not popular.)

Cleaner `SPEC` data structures:
* (which is the main motivation of this change.)
* Values are simple lists, instead of individual `validator` objects, so the data structure can easily be dumped out (or even loaded in). Each list looks like `[value_type, default, allowed_2, allowed_3, ...]` which should be enough to represent the essentials, where:
  * `value_type`: is the compulsory value type of the setting, and defined only using constants from `parsec.validate.ParsecValidator`.
  * `default`: is the optional default value. Default is `None` if this is not defined.
  * `allowed_2, ...`: are the only other allowed values of this setting, if specified.

Other changes:
* Single validator with simple dispatch table for *coercers* methods. All *coercers* moved to live under `parsec.validate.ParsecValidator`.
* Clearer separation of concerns between `parsec.config` and `parsec.validate`, and between `parsec.*` and `cylc.cfgspec.*`.
* `cylc.cfgspec.*` modules no longer export the singleton configuration objects - so configuration files should only be loaded on demand at run time, instead of at compile time.

hjoliver commented:
This seems generally sensible.  (I'm not sure that on-demand loading of config files is much of an advantage in our context - suite configs are the only big ones, and they're always "demanded" more or less immediately, no? - but that's not really an objection).

matthewrmshin commented:
Quick answers:

The change is mainly about simplification, although I was looking at ways to improve the data structure of the configuration objects and the data structure of their specifications.

On singleton. We have already eliminated the `suite.rc` singleton in #2373, and we have loading on demand of `global.rc` in #2601. This change will apply to the remaining, mainly the GUI configuration files, so this is really for completeness rather than a necessity.

On loading/dumping. I am really talking about the specification data structure, instead of the configuration here. It may be good to be able to load the specification on demand - to reduce the memory footprint of loading the `cylc.cfgspec.*` modules.

oliver-sanders commented on lib/parsec/validate.py:
```diff
@@ -23,365 +23,511 @@
+            raise IllegalValueError('boolean', keys, value)
+
+    @classmethod
+    def _coerce_cycle_point(cls, value, keys):
```
Are we able to use the `cylc.cycling.iso8601.SuiteSpecifics.iso8601_parsers[0]` instance of `TimePointParser` rather than creating a new one here or is that not available at this stage?

matthewrmshin replied:
It is not yet available at this stage, and will introduce a circular dependency between parsec and cylc libraries. Having said, I'll take a look to see whether it is possible to do it the other way round or not.

matthewrmshin replied:
(Having looked at it, it may be more trouble than it is worth.)

hjoliver commented:
(We should get this one in once you've resolved the current conflicts)

matthewrmshin commented:
Quite difficult conflicts, but I think I've got something working. There will be some renames and relocations of classes. (I'm on a course today and tomorrow. I'll reassess this as soon as I'm back on Friday.)

matthewrmshin commented:
This is good enough again. (Conflict resolved. Tests pass.) I have done some relocation and renames to ensure that the `parsec` package can still be **independent** of the `cylc` package. In particular:
* `cylc.wallclock` is now just `wallclock`. I have also removed the `cylc.flags.utc` flag in favour of an encapsulated global setting in `wallclock`.
* The context classes in `cylc.mp_pool` are moved to `parsec.validate`. They are renamed `SubProcContext` and `SubFuncContext`. (Not the best natural home for these classes, so please feel free to suggest a better location.)

hjoliver commented:
> The context classes in cylc.mp_pool are moved to parsec.validate. They are renamed SubProcContext and SubFuncContext. (Not the best natural home for these classes, so please feel free to suggest a better location.)

Yeah, I don't think `lib/parsec` is the best place for these, or for non-generic "coercer" functions (where by generic I mean lists of basic types etc.) - as none of these entities mean anything to a generic config file parser.  Could the `SubXXXContext`s go back to `lib/cylc`, and the non-generic coercers be implemented in a subclass of `ParsecValidator` in `lib/cylc/cfgspec`?   (Maybe none of this is worth the hassle though, as in reality parsec is not likely to ever be used outside of Cylc...).

matthewrmshin commented:
Fair enough. I'll think about a better home for them.

matthewrmshin commented:
> (Maybe none of this is worth the hassle though, as in reality parsec is not likely to ever be used outside of Cylc...)

I have also considered this one as part of this change, but have decided to leave it as-is for now. A different approach would be to bring `parsec` into Cylc with e.g. the `cylc.parsec` name space.

hjoliver commented:
> A different approach would be to bring parsec into Cylc with e.g. the cylc.parsec name space.

That seems sensible. (Note I had originally intended to re-use `parsec` in an in-house Python program that uses the ancient ConfigObj, hence the separation from `lib/cylc`, but that ain't gonna happen at this point.)

matthewrmshin commented:
Updated and re-implemented validator via sub-class. I have moved around a few things, but I hope the name-spaces are now making sense again. (See log message of latest commit.)

matthewrmshin commented:
Conflicts resolved. Should be ready to review again.

hjoliver commented:
(next top priority!)

hjoliver commented on lib/parsec/config.py:
```diff
@@ -35,29 +35,32 @@ def __init__(self, msg):
+    def checkspec(spec_root, parents=None):
+        """Check that the file spec is a nested dict of specifications"""
+        stack = [[spec_root, []]]
+        while stack:
```
To my mind, recursing through a nested dict is simpler as well as shorter than iteration, and (somewhat surprisingly!) it seems to be faster too. I think I asked you before (when your replaced another of my recursive implementations!) why you prefer iteration, but I can't recall the answer - can you remind me?

matthewrmshin replied:
Sorry it is really an educated preference. Recursion is easier to read but a bit more difficult to understand, at least for my brain!

There are several reasons for the preference. One is that iteration is easier to parallelize than recursion - should there be a necessity to do so. The other reason is my previous life being a Perl developer, where you can get warning for deep recursion even though the code is perfectly good. Deep recursion can be costly in memory as well - due to the function stack.

None applicable here, I guess. Happy to revert if you feel strongly about it.

hjoliver replied:
No, that's fine - I don't feel strongly enough about it to force a revert.

matthewrmshin commented:
@oliver-sanders please sanity check.

matthewrmshin commented:
Resolved conflict by rebasing.

matthewrmshin commented:
@oliver-sanders please prioritise review of this one.
